### PR TITLE
#2694 - Student PT life time loan balance - Part 1

### PR DESCRIPTION
**Pull request by dheepak-aot**
## Student Part Time loan balance display - Part 1

- [x] Renamed the `Overawards` tab to `Balances` and related routes and components in ministry student search and institution student search.
image
image

- [x] Created Ministry API for student part time monthly loan balance. Values returned from API are static now.
- [x] Created a Vue component for ministry to display the student part time loan balances. Added the new component to Balances view in ministry student search tab.

image

- [x] Created Api and Service at Vue

## To be covered in upcoming PR
- [ ] Backend service and implementation of student loan balance.
- [ ] E2E tests.

### CODE REVIEW

File: sources/packages/web/src/views/institution/student/InstitutionStudentBalances.vue
Comment (sh16011993) [inline sources/packages/web/src/views/institution/student/InstitutionStudentBalances.vue:1]: Shouldn't this file contain the component: `<student-loan-balance-part-time :student-id="studentId" />`

I don't see the balances tables at the bottom of the balances tab in the intuition login.

File: sources/packages/backend/apps/api/src/route-controllers/student-loan-balance/models/student-loan-balance.dto.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+/**
+ * Student loan balance details of a month.
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/route-controllers/student-loan-balance/models/student-loan-balance.dto.ts:2]: As a suggestion, the word monthly can be misleading and I am not sure if adds some value. The report is received monthly but the balance "is just a balance" that represents the lastest state.

File: sources/packages/backend/apps/api/src/route-controllers/student-loan-balance/models/student-loan-balance.dto.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+ */
+class StudentMonthlyLoanBalanceDetailAPIOutDTO {
+  balanceDate: string;
+  cslBalance: number;
+}
+
+/**
+ * Monthly loan balance details of a student
+ * upto 12 months from current date.
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/route-controllers/student-loan-balance/models/student-loan-balance.dto.ts:11]: Small type in "upto".

Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/route-controllers/student-loan-balance/models/student-loan-balance.dto.ts:11]: Minor, not a blocker, I am not sure if "upto 12 months from current date." is meaningful to be added at DTO level.

File: sources/packages/backend/apps/api/src/route-controllers/student-loan-balance/student-loan-balance.aest.controller.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+@Controller("student-loan-balance")
+@ApiTags(`${ClientTypeBaseRoute.AEST}-student-loan-balance`)
+export class StudentLoanBalanceAESTController extends BaseController {
+  constructor(private readonly studentService: StudentService) {
+    super();
+  }
+
+  /**
+   * Get the monthly part-time loan balance details of
+   * the given student upto 12 months.
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/route-controllers/student-loan-balance/student-loan-balance.aest.controller.ts:28]: Small typo in "upto".

File: sources/packages/backend/apps/api/src/route-controllers/student-loan-balance/student-loan-balance.aest.controller.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+   * @param studentId student.
+   * @returns student monthly part-time loan balance.
+   */
+  @ApiNotFoundResponse({
+    description: "Student not found.",
+  })
+  @Get("part-time/student/:studentId")
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/route-controllers/student-loan-balance/student-loan-balance.aest.controller.ts:35]: I am not sure about using the `part-time` at this route and at this level. Maybe it can be a future filter but right now we are not differentiating anywhere else.

File: sources/packages/backend/apps/api/src/route-controllers/student-loan-balance/student-loan-balance.aest.controller.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+  async getStudentPartTimeLoanBalance(
+    @Param("studentId", ParseIntPipe) studentId: number,
+  ): Promise<StudentLoanBalanceAPIOutDTO> {
+    const studentExist = await this.studentService.studentExists(studentId);
+    if (!studentExist) {
+      throw new NotFoundException("Student not found.");
+    }
+    //TODO: Remove the static values when actual implementation is done.
+    const staticData: StudentLoanBalanceAPIOutDTO = {
+      loanBalanceDetails: [
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/route-controllers/student-loan-balance/student-loan-balance.aest.controller.ts:45]: I know that it is just mocked data but please ensure the `desc` order for the final data returned.

File: sources/packages/web/src/components/aest/students/StudentLoanBalancePartTime.vue
Comment (andrewsignori-aot) [inline sources/packages/web/src/components/aest/students/StudentLoanBalancePartTime.vue:1]: Similar to the below one, I would refrain from using `part-time`.
https://github.com/bcgov/SIMS/pull/3038/files#r1539949595

File: sims.code-workspace [hunk 6]
```diff
@@ -98,7 +98,8 @@
       "SABC",
       "SFAS",
       "timestamptz",
-      "typeorm"
+      "typeorm",
```
Comment (guru-aot) [inline sims.code-workspace:101]: 👍

File: sources/packages/backend/apps/api/src/route-controllers/student-loan-balance/models/student-loan-balance.dto.ts [hunk 7]
```diff
@@ -0,0 +1,15 @@
+/**
+ * Student loan balance details of a month.
+ */
+class StudentMonthlyLoanBalanceDetailAPIOutDTO {
```
Comment (guru-aot) [inline sources/packages/backend/apps/api/src/route-controllers/student-loan-balance/models/student-loan-balance.dto.ts:4]: Can we have the Monthly removed from the class name?

**Review by andrewsignori-aot — COMMENTED** [review]
Nice work, please take a look at the comments.

File: sources/packages/web/src/components/aest/students/StudentLoanBalancePartTime.vue [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+      const studentMonthlyLoanBalance =
+        await StudentLoanBalanceService.shared.getStudentPartTimeLoanBalance(
+          props.studentId,
+        );
+      studentLoanBalanceDetails.value =
+        studentMonthlyLoanBalance.loanBalanceDetails;
+    });
+
+    return {
+      page,
```
Comment (sh16011993) [inline sources/packages/web/src/components/aest/students/StudentLoanBalancePartTime.vue:75]: I don't see `page` being used in vue?

File: sources/packages/web/src/services/StudentLoanBalanceService.ts [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+  // Shared Instance
+  private static instance: StudentLoanBalanceService;
+
+  static get shared(): StudentLoanBalanceService {
+    return this.instance || (this.instance = new this());
+  }
+
+  /**
+   * Get the monthly part-time loan balance details of
+   * the given student upto 12 months.
```
Comment (sh16011993) [inline sources/packages/web/src/services/StudentLoanBalanceService.ts:17]: Isn't it better to put the comment as something like `retrieving the last 12 records` rather than the records for the last 12 months?

File: sources/packages/web/src/components/aest/students/StudentLoanBalancePartTime.vue [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+export default defineComponent({
+  props: {
+    studentId: {
+      type: Number,
+      required: true,
+    },
+  },
+  setup(props) {
+    const page = ref(DEFAULT_PAGE_NUMBER);
+    const pageLimit = ref(DEFAULT_PAGE_LIMIT);
```
Comment (guru-aot) [inline sources/packages/web/src/components/aest/students/StudentLoanBalancePartTime.vue:59]: the DEFAULT_PAGE_LIMIT value is 10 and we are planning to show only the last 12 records, is that fine to have 10 as a limit and 2 in the next page or show all 12? also as per the screenshot we show only 5 values in the page, is that the right limit?

File: sources/packages/web/src/constants/routes/RouteConstants.ts [hunk 11]
```diff
@@ -79,7 +79,7 @@ export const InstitutionRoutesConst = {
   STUDENT_APPLICATION_DETAILS: Symbol(),
   STUDENT_RESTRICTIONS: Symbol(),
   STUDENT_FILE_UPLOADS: Symbol(),
-  STUDENT_OVERAWARDS: Symbol(),
+  STUDENT_BALANCES: Symbol(),
```
Comment (guru-aot) [inline sources/packages/web/src/constants/routes/RouteConstants.ts:82]: In my APi i have used Student loan balances, just to be more understandable. Not a blocker on the naming, but just a suggestion.

**Review by sh16011993 — COMMENTED** [review]
Nice Work @dheepak-aot 👍

File: sources/packages/web/src/components/aest/students/StudentLoanBalancePartTime.vue (hunk 8 above)
Comment (dheepak-aot) [reply sources/packages/web/src/components/aest/students/StudentLoanBalancePartTime.vue:75]: Nice catch.

File: sources/packages/web/src/components/aest/students/StudentLoanBalancePartTime.vue (hunk 10 above)
Comment (dheepak-aot) [reply sources/packages/web/src/components/aest/students/StudentLoanBalancePartTime.vue:59]: I would say, unless it is very necessary we shouldn't modify the default behavior.

File: sources/packages/web/src/constants/routes/RouteConstants.ts (hunk 11 above)
Comment (dheepak-aot) [reply sources/packages/web/src/constants/routes/RouteConstants.ts:82]: This route is corresponding to the tab which was previously called overawards and now called balances.

image

File: sources/packages/backend/apps/api/src/route-controllers/student-loan-balance/student-loan-balance.aest.controller.ts (hunk 4 above)
Comment (dheepak-aot) [reply sources/packages/backend/apps/api/src/route-controllers/student-loan-balance/student-loan-balance.aest.controller.ts:35]: I added the `part-time` to the API (and also to the component) as we are precisely dealing with part-time data here.

I see the point of API being more generic with respect to student loan balance. But from the UI perspective the Vue component is about part time loan balance and I prefer that to be named with part time. Let me know your thoughts.

Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/route-controllers/student-loan-balance/student-loan-balance.aest.controller.ts]: I am ok with the way you described 😉

File: sources/packages/backend/apps/api/src/route-controllers/student-loan-balance/student-loan-balance.aest.controller.ts [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+@Controller("student-loan-balance")
+@ApiTags(`${ClientTypeBaseRoute.AEST}-student-loan-balance`)
+export class StudentLoanBalanceAESTController extends BaseController {
+  constructor(private readonly studentService: StudentService) {
+    super();
+  }
+
+  /**
+   * Get loan balance details of
+   * the given student upto 12 recent records received.
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/route-controllers/student-loan-balance/student-loan-balance.aest.controller.ts:28]: Typo in "upto".

File: sources/packages/backend/apps/api/src/route-controllers/student-loan-balance/student-loan-balance.aest.controller.ts [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+  Controller,
+  Get,
+  NotFoundException,
+  Param,
+  ParseIntPipe,
+} from "@nestjs/common";
+import { ApiNotFoundResponse, ApiTags } from "@nestjs/swagger";
+import { AuthorizedParties } from "../../auth/authorized-parties.enum";
+import { AllowAuthorizedParty, Groups } from "../../auth/decorators";
+import { UserGroups } from "../../auth/user-groups.enum";
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/route-controllers/student-loan-balance/student-loan-balance.aest.controller.ts:11]: Pretty minor and not a blocker, and maybe not part of the PR, we introduced an index under auth and we never actually started to use it.
All auth-related imports can be under one single index.

File: sources/packages/web/src/services/StudentLoanBalanceService.ts [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+  // Shared Instance
+  private static instance: StudentLoanBalanceService;
+
+  static get shared(): StudentLoanBalanceService {
+    return this.instance || (this.instance = new this());
+  }
+
+  /**
+   * Get loan balance details of
+   * the given student upto 12 recent records received.
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/services/StudentLoanBalanceService.ts:17]: Typo on "upto".

**Review by andrewsignori-aot — APPROVED** [review]
Thanks for making the changes, only optional minor comments left. Looks good 👍

**Review by sh16011993 — APPROVED** [review]
Nice Work @dheepak-aot 👍  LGTM. Thanks for making the changes.

**Review by guru-aot — APPROVED** [review]
LGTM, nice work @dheepak-aot

File: sources/packages/backend/apps/api/src/route-controllers/student-loan-balance/student-loan-balance.aest.controller.ts (hunk 13 above)
Comment (dheepak-aot) [reply sources/packages/backend/apps/api/src/route-controllers/student-loan-balance/student-loan-balance.aest.controller.ts:11]: Sure I will do this change in the upcoming PR.
